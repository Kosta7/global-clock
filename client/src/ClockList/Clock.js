import React from "react";
import styled from "styled-components";
import { format, getMinutes, setMinutes, startOfMinute } from "date-fns";
import PropTypes from "prop-types";
import * as d3 from "d3";

import getTimeScale from "./getTimeScale";
import { timeline, colors } from "ui/constants";

import { DeleteClockButton } from "./DeleteClockButton";
import { TimePicker } from "./TimePicker";

export class Clock extends React.Component {
  state = {
    isPickDateMode: false
  };

  componentDidMount() {
    if (!this.props.isEditMode) {
      this.drawTimeChart();
    }
    window.addEventListener("resize", this.onWindowResize.bind(this));
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.time.getTime() !== this.props.time.getTime() &&
      !this.props.isEditMode
    ) {
      this.clearTimeChart();
      this.drawTimeChart();
    } else if (!prevProps.isEditMode && this.props.isEditMode) {
      this.clearTimeChart();
    } else if (prevProps.isEditMode && !this.props.isEditMode) {
      this.drawTimeChart();
    }
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.onWindowResize.bind(this));
  }

  render() {
    const scrolledTime = getScrolledTime(this.props.time, this.props.shift);

    let roundedTime =
      this.props.shift === 0
        ? scrolledTime
        : roundToNearestMinutes(scrolledTime, 15);

    const formattedTime = format(roundedTime, "hh:mm A");

    const timezoneInHours = `${this.props.timezoneOffset > 0 ? "+" : ""}${this
      .props.timezoneOffset / 60}`;

    return (
      <StyledClock ref={el => (this.clockRef = el)}>
        <City>
          {this.props.city}
          {this.props.isEditMode && ` (${timezoneInHours})`}
        </City>
        {!this.props.isEditMode && <Time>{formattedTime}</Time>}

        {!this.props.isEditMode && (
          <PickDateButton onClick={this.togglePickDateMode.bind(this)}>
            <span role="img" aria-label="pick date">
              ðŸ“…
            </span>
          </PickDateButton>
        )}

        {this.props.isEditMode && (
          <DeleteClockButton onClick={this.deleteClockHandler.bind(this)} />
        )}

        {this.state.isPickDateMode && (
          <TimePicker
            city={this.props.city}
            onSubmit={this.pickTime.bind(this)}
            closePopup={this.togglePickDateMode.bind(this)}
          />
        )}
      </StyledClock>
    );
  }

  drawTimeChart() {
    const timeScale = getTimeScale(this.props.time);

    const xAxis = d3
      .axisBottom(timeScale)
      .ticks(d3.timeDay.every(1))
      .tickSizeOuter(0);

    const svg = d3
      .select(this.clockRef)
      .append("svg")
      .attr("width", window.innerWidth * timeline.width)
      .attr("height", 50);

    svg
      .append("g")
      .call(xAxis)
      .style("transform", "translateY(25px)")
      .style("color", colors.text)
      .style("font-size", "12px")
      .style("font-family", "monospace")
      .style("user-select", "none");
  }

  clearTimeChart() {
    d3.select(this.clockRef)
      .selectAll("svg")
      .remove();
  }

  deleteClockHandler() {
    this.props.deleteClock(this.props.id);
  }

  togglePickDateMode() {
    this.setState({ isPickDateMode: !this.state.isPickDateMode });
  }

  pickTime(time) {
    this.props.pickTime(time, this.props.timezoneOffset);
  }

  onWindowResize() {
    this.clearTimeChart();
    if (!this.props.isEditMode) {
      this.drawTimeChart();
    }
  }
}
Clock.propTypes = {
  city: PropTypes.string.isRequired,
  time: PropTypes.instanceOf(Date).isRequired,
  shift: PropTypes.number.isRequired,
  timezoneOffset: PropTypes.number.isRequired,
  isisEditMode: PropTypes.bool,
  deleteClock: PropTypes.func
};

function roundToNearestMinutes(date, interval) {
  var roundedMinutes = Math.floor(getMinutes(date) / interval) * interval;
  return setMinutes(startOfMinute(date), roundedMinutes);
}

function getScrolledTime(time, shift) {
  return new Date(Number(time) + shift);
}

const Time = styled.div`
  font-size: 22px;
  font-weight: 800;
  color: rgba(255, 255, 255, 80%);
  font-family: monospace;
  display: inline-flex;
  justify-content: space-between;
  position: absolute;
  left: calc(60% + 5px);
  transform: translateY(-2px);
  pointer-events: none;
`;

const City = styled.div`
  position: absolute;
  left: 5px;
  font-family: monospace;
  font-size: 18px;
  color: ${colors.text};
  pointer-events: none;
`;

const PickDateButton = styled.div`
  opacity: 0;
  position: absolute;
  right: 5px;
  cursor: pointer;
  pointer-events: none;
  transition: opacity 0.2s;
`;

const StyledClock = styled.div`
  height: 50px;
  margin-bottom: 25px;
  width: 100%;

  &:hover ${PickDateButton} {
    opacity: 1;
    pointer-events: auto;
  }
`;
