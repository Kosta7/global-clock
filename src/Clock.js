import React from "react";
import styled from "styled-components";
import {
  format,
  isToday,
  isYesterday,
  isTomorrow,
  addHours,
  addMinutes
} from "date-fns";

const Clock = ({ city, timezone, shift, className }) => {
  const date = timezone ? getRemoteDate(timezone, shift) : getLocalDate(shift);
  const formattedDate = format(date, "hh:mm A");

  const differentDay = isToday(date)
    ? ""
    : isYesterday(date)
      ? "Yesterday"
      : isTomorrow(date)
        ? "Tomorrow"
        : format(date, "dddd");

  return (
    <div className={className}>
      <Time>
        {formattedDate}
        <DifferentDay>{differentDay}</DifferentDay>
      </Time>
      <City>{city}</City>
    </div>
  );
};

const getLocalDate = shift => addHours(new Date(), shift);

const getRemoteDate = (timezone, shift) => {
  const localDate = getLocalDate(shift);
  const localTimezone = localDate.getTimezoneOffset();

  const remoteTimezone = -timezone;

  const timezoneDifference = localTimezone - remoteTimezone;

  return addMinutes(localDate, timezoneDifference);
};

const Time = styled.div`
  font-size: 30px;
  font-weight: 600;
  color: #ffb432;
  font-family: monospace;
  display: inline-flex;
  justify-content: space-between;
  width: 100%;
`;

const DifferentDay = styled.span`
  font-size: 22px;
  font-weight: 400;
`;

const City = styled.div`
  font-size: 20px;
  color: #737dc3;
`;

const StyledClock = styled(Clock)`
  display: flex;
  flex-direction: column;
  padding: 10px 20px;
  margin: 5px;
  border: 1px solid #b4c8d7;
  border-radius: 3px;
`;

export { StyledClock as Clock };
