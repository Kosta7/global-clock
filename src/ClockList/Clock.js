import React from "react";
import styled from "styled-components";
import { format } from "date-fns";
import PropTypes from "prop-types";

import { DeleteClockButton } from "./DeleteClockButton";

const Clock = ({
  id,
  city,
  date,
  timezone,
  editMode,
  deleteClock,
  className
}) => {
  const formattedDate = format(date, "hh:mm A");

  const deleteClockHandler = () => deleteClock(id);

  return (
    <div className={className}>
      <City>{city}</City>
      <Time>{formattedDate}</Time>
      {editMode && <DeleteClockButton deleteClock={deleteClockHandler} />}
    </div>
  );
};
Clock.propTypes = {
  city: PropTypes.string.isRequired,
  timezone: PropTypes.number,
  shift: PropTypes.number.isRequired,
  editMode: PropTypes.bool,
  deleteClock: PropTypes.func
};

const Time = styled.div`
  font-size: 18px;
  font-weight: 600;
  color: #ffb432;
  font-family: monospace;
  display: inline-flex;
  justify-content: space-between;
  position: absolute;
  left: calc(50% + 5px);
  pointer-events: none;
`;

const City = styled.div`
  position: absolute;
  left: 5px;
  font-family: monospace;
  font-size: 18px;
  color: #737dc3;
  pointer-events: none;
`;

const StyledClock = styled(Clock)`
  min-height: 25px;
  margin-bottom: 50px;
  width: 100%;
  border-bottom: 1px dashed #b4c8d7;
`;

export { StyledClock as Clock };
