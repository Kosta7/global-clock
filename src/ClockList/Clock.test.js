import { Clock } from "./Clock";

it("renders correctly", () => {
  const wrapper = shallow(<Clock city={`New York`} timezone={-7} shift={2} />);

  expect(wrapper).toMatchSnapshot();
});

it("renders correctly again", () => {
  const wrapper = render(<Clock city={`New York`} timezone={-7} shift={2} />);
  expect(wrapper).toMatchSnapshot();
});

// it("formats city correctly", () => {
//   const wrapper = mount(<Clock city={`New York`} timezone={-7} shift={2} />);

//   // set property displayName of City component to "City" so that enzyme could find it
//   const text = wrapper.find("City").text();
//   expect(text).toEqual("New York");
// });
