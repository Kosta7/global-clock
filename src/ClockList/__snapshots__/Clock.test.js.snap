// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<StyledComponent
  city="New York"
  forwardedClass={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-ifAKCX",
        "isStatic": true,
        "rules": Array [
          "
  position: relative;
  display: flex;
  flex-direction: column;
  padding: 10px 20px;
  margin-bottom: 5px;
  border: 1px solid #b4c8d7;
  border-radius: 3px;
",
        ],
      },
      "displayName": "Styled(Clock)",
      "render": [Function],
      "styledComponentId": "sc-ifAKCX",
      "target": [Function],
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
      Symbol(Symbol.hasInstance): [Function],
    }
  }
  forwardedRef={null}
  shift={2}
  timezone={-7}
/>
`;

exports[`renders correctly again 1`] = `
<div
  class="sc-ifAKCX dmSHNx"
>
  <div
    class="sc-bwzfXH gmmHMg"
  >
    11:04 PM
    <span
      class="sc-htpNat fbWjVY"
    >
      Yesterday
    </span>
  </div>
  <div
    class="sc-bxivhb iWzsho"
  >
    New York
  </div>
</div>
`;
